docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+
+This storage plugin will respond "yes" when asked if it owns a volume spec that contains a PV + `sticky-host` annotation, and "no" in every other case. (TODO: do we need to bind mount the directory for every pod?).
+
+## Example
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: test
+  annotations:
+    # Set by scheduler or admin
+    volume.alpha.kubernetes.io/sticky-host: ""
+spec:
+  capacity:
+    storage: 10Mi
+  accessModes:
+    - ReadWriteMany
+  emptyDir: {}
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test
+  annotations:
+    # Mentioned for the sake of clarity
+    volume.alpha.kubernetes.io/storage-class: free
+spec:
+  accessModes: [ "ReadWriteMany" ]
+  resources:
+    requests:
+      storage: 10Mi
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: test-
+spec:
+  terminationGracePeriodSeconds: 0
+  containers:
+  - name: busybox
+    image: busybox
+    command:
+    - sh
+    - -c
+    - touch /data/`hostname`; while true; do sleep 3600; done
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  volumes:
+  - name: data
+    persistentVolumeClaim:
+      claimName: test
+```
+
+Creating this will give you a pod pinned to the node the scheduler initally picks, with a directory mounted at /data backed by `/var/lib/kubelet/volumes/pv-uuid`. Creating the same thing without the PV section will invoke a dynamic storage provisioner to create an identical PV.
+
+## Alternatives
+
+I (@bprashanth) belive that a sticky emptydir volume as proposed strikes a good balance between hostPath local storage and a spectrum of other possibilities such as:
+* Hostpath dynamic provisioner: The petset team is still considering a short term solution that involves a dynamic provisioner creating a PV backed by hostPath, pointing to the same `/var/lib/kubelet/volumes/pv-uuid` location. This requires 0 code modifications in the node volume subsystem, but requires the scheduler and kubelet admission time changes (relatively simple). This also means the sticky volumes will not get cleaned up, so it is not a viable long term plan.
+* Streamline distributed filesystem deployment on Kubernetes: If we handwave the performance benefits of using local storage (writing to local disk is ~ going over a LAN), we could in theory make it really easy to deploy eg: Gluster and leverage local storage. While this is still a possibility, it introduces another failure mode, and distributed filesystems are far from being easy to deploy. It also significantly raises the bar for the dev/test use case.
+* Use an NFS RC: This is already possible, but still doesn't give us the performance benefit, and is tedious to setup (though not as tedious as a distributed filesystem).
```

Comment:
It'll be a storage classes thing.  Now that storage classes proposals are
approved and moving forward, local storage is just another class of storage.

On Mon, Aug 15, 2016 at 5:31 AM, keyingliu <notifications@github.com> wrote:

> In docs/proposals/emptydir-pv.md
> <https://github.com/kubernetes/kubernetes/pull/30044#discussion_r74738800>
> :
>
> > +    - name: data
> > +      mountPath: /data
> > +  volumes:
> > +  - name: data
> > +    persistentVolumeClaim:
> > +      claimName: test
> > +```
> > +
> > +Creating this will give you a pod pinned to the node the scheduler initally picks, with a directory mounted at /data backed by `/var/lib/kubelet/volumes/pv-uuid`. Creating the same thing without the PV section will invoke a dynamic storage provisioner to create an identical PV.
> > +
> > +## Alternatives
> > +
> > +I (@bprashanth) belive that a sticky emptydir volume as proposed strikes a good balance between hostPath local storage and a spectrum of other possibilities such as:
> > +* Hostpath dynamic provisioner: The petset team is still considering a short term solution that involves a dynamic provisioner creating a PV backed by hostPath, pointing to the same `/var/lib/kubelet/volumes/pv-uuid` location. This requires 0 code modifications in the node volume subsystem, but requires the scheduler and kubelet admission time changes (relatively simple). This also means the sticky volumes will not get cleaned up, so it is not a viable long term plan.
> > +* Streamline distributed filesystem deployment on Kubernetes: If we handwave the performance benefits of using local storage (writing to local disk is ~ going over a LAN), we could in theory make it really easy to deploy eg: Gluster and leverage local storage. While this is still a possibility, it introduces another failure mode, and distributed filesystems are far from being easy to deploy. It also significantly raises the bar for the dev/test use case.
> > +* Use an NFS RC: This is already possible, but still doesn't give us the performance benefit, and is tedious to setup (though not as tedious as a distributed filesystem).
>
> A question about PV/PVC bind: how can you make sure the emptyDir PVC is
> bound to emptyDir PV rather than other PVs (NFS, Cinder)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30044/files/2a9e5a2908a31832129d7cab502b76b41b016069#r74738800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5CfFB2jze3S9ptd3H8tqOl7OFH8ks5qgDH6gaJpZM4JcSCY>
> .
>